### DISCUSSION TITLE
Specifying types for extra fields

### DISCUSSION THREAD

**Discussion by mdekstrand** [pydantic/pydantic#9442] [Question] [answered]
We have a data modeling problem where we have a model that has one known field of type `A`, and unknown fields all of type `B`. Effectively, it is a `dict[str,B]`, except one name `'foo'` has type `A`. We'd like to be able to parse this into something like this:

```py
class Container(BaseModel):
    foo: A
    others: dict[str, B]
```

Not wedded to that specific class design, but it's the basic idea. I know approximately how I would write this in Zod:

```js
zod.object({
    foo: A
)}.intersection(zod.record(zod.string(), B))
```

What is a good way to represent and validate such a model in Pydantic?

**Comment by x11x** [comment]
One way could be to use the ["extra fields" functionality](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.extra), and a [TypeAdapter](https://docs.pydantic.dev/latest/concepts/type_adapter/) to validate the extra data. There are probably many variations on this pattern, and possible improvements to my code, but here is a rough example

```python
from typing import Self
from pydantic import BaseModel, Field, TypeAdapter, ValidationInfo, model_validator


class A(BaseModel):
    ...

class B(BaseModel):
    ...


others_type_adapter = TypeAdapter(dict[str, B])


class Container(BaseModel, extra='allow'):
    foo: A
    others: dict[str, B] = Field(init=False, default_factory=dict)

    @model_validator(mode='after')
    def _validate_extras(self, info: ValidationInfo) -> Self:
        extra = self.model_extra
        if extra:
            # TODO should catch ValidationError and wrap it so it appears like any other field validation error
            others = others_type_adapter.validate_python(extra, context=info.context)
            extra.clear()  # (otherwise data would be stored twice)
            self.others = others
        return self
```

A few issues with this:
  - To give consistent, useful error messages, I think it is necessary to catch the `ValidationError` from `others_type_adapter.validate_python` and wrap it to include information to indicate that it came from "others" input, etc. I haven't worked out exactly how to do this but try https://docs.pydantic.dev/latest/api/pydantic_core/#pydantic_core.InitErrorDetails and other related error classes
  - JSON mode vs Python mode -- I am validating the extra data in "Python" mode. I'm pretty sure this is correct, because even in JSON mode, the values will have been parsed into the basic JSON-compatible types (int/float/bool/str/None and list/dict of these). I'm just not sure if in every case (1) _validating in Python mode on data parsed from JSON_ is the same as (2) _validating in JSON mode right from the start_, but I *think* Pydantic is designed to make these two ways of parsing equivalent.
    - Specific things to check would be datetimes etc, or types that are serialized to strings in the JSON.
    -  If you need, you can do different validation logic in `_validate_extras` by detecting `info.mode` (which will be either "python" or "json")
    -  It would be important to test with actual data and identify how different types parse in the different modes; i.e. take just the `others` data (with `foo` removed), and see how it parses using `others_type_adapter.validate_python` vs `others_type_adapter.validate_json` vs `others_type_adapter.validate_strings`.
    - Also, it could be helpful to refer to the [Conversion Table](https://docs.pydantic.dev/latest/concepts/conversion_table/)
     - there is also [BaseModel.model_validate_strings](https://docs.pydantic.dev/latest/concepts/models/#helper-functions) which would seem to be useful for the case of `info.mode == "JSON"`, but unfortunately I don't think this is applicable here because it requires "all" values to be strings (as I said they will be parsed into int/bool/float etc/).

**Comment by Viicos** [comment, marked as answer]
You are looking for https://docs.pydantic.dev/latest/concepts/models/#extra-fields. In particular, the last section about annotating `__pydantic_extra__` to provide additional validation.

**Reply by x11x** [reply]
Thanks!
I read that but the the wording made me think extras still wouldn't be validated -- but closer reading and the example does clearly show they are. My mistake. Makes things simpler!
